File: python/linkedlist_test.py
import unittest
from linkedlist import LinkedList

values = [
    {'val': 2},
    {'val': 4},
    {'val': 6},
    {'val': 8},
    {'val': 10},
]

class TestLinkedList(unittest.TestCase):
    def test_list_add(self):
        global values
        linkedlist = LinkedList()

        for i, v in enumerate(values):
            linkedlist.add(v)
            self.assertEqual(linkedlist.size(), i+1)
        
        for i, v in enumerate(values):
            val = linkedlist.get(i)
            self.assertEqual(val['val'], v['val'])


    def test_list_get_from_array(self):
        global values
        linkedlist = LinkedList(values)

        for idx, val in enumerate(values):
            v = linkedlist.get(idx)
            self.assertEqual(v['val'], val['val'])


    def test_list_remove_element(self):
        global values
        linkedlist = LinkedList(values)

        last = linkedlist.remove(4)
        self.assertEqual(last['val'], 10)
        self.assertEqual(linkedlist.size(), 4)

        middle = linkedlist.remove(2)
        self.assertEqual(middle['val'], 6)
        self.assertEqual(linkedlist.size(), 3)

        first = linkedlist.remove(0)
        self.assertEqual(first['val'], 2)
        self.assertEqual(linkedlist.size(), 2)


    def test_list_to_string(self):
        global values
        linkedlist = LinkedList(values)
        s = linkedlist.to_string(lambda x: str(x['val']))
        self.assertEqual(s, '[2,4,6,8,10]')


if __name__ == '__main__':
    unittest.main()

File: java/src/test/java/com/example/LinkedListTest.java
package com.example;

import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestSuite;

public class LinkedListTest extends TestCase {
    public LinkedListTest(String testName) {
        super(testName);
    }

    public static Test suite() {
        return new TestSuite(LinkedListTest.class);
    }

    private final Struct[] values = {
            new Struct(2),
            new Struct(4),
            new Struct(6),
            new Struct(8),
            new Struct(10),
    };

    public void testListAdd() {
        LinkedList<Struct> list = new LinkedList<>();

        for (int i = 0; i < values.length; i++) {
            Struct val = values[i];
            list.add(val);

            assertEquals(list.size(), i + 1);
        }

        for (int i = 0; i < values.length; i++) {
            Struct val = values[i];
            Struct v = list.get(i).get();
            assertEquals(val.val, v.val);
        }
    }

    public void testListGetFromArray() {
        LinkedList<Struct> list = new LinkedList<>(values);

        for (int i = 0; i < values.length; i++) {
            Struct val = values[i];
            Struct v = list.get(i).get();
            assertEquals(val.val, v.val);
        }
    }

    public void testListRemoveElement() {
        LinkedList<Struct> list = new LinkedList<>(values);

        Struct last = list.remove(4).get();
        assertEquals(last.val, 10);
        assertEquals(list.size(), 4);

        Struct middle = list.remove(2).get();
        assertEquals(middle.val, 6);
        assertEquals(list.size(), 3);

        Struct first = list.remove(0).get();
        assertEquals(first.val, 2);
        assertEquals(list.size(), 2);
    }

    public void testListToString() {
        LinkedList<Struct> list = new LinkedList<>(values);
        String str = list.toString((Struct s) -> String.valueOf(s.val));
        String expected = "[2,4,6,8,10]";
        assertEquals(str, expected);
    }
}

class Struct {
    public int val;

    public Struct(int value) {
        this.val = value;
    }
}